listenbrainz/listenstore/timescale_fill_userid.py

```diff
@@ -0,0 +1,144 @@
+    if missing_users:
+        # If we have listens with a username that isn't in the user table, report it but only
+        # if we haven't seen these usernames before
+        if missing_users - global_missing_users:
+            global_missing_users.update(missing_users)
+            print(f" - oops - found some usernames who have listens but no user row")
+            print(f"   all missing users are now {global_missing_users}")
+
+    if not user_mapping:
+        print(" - Unexpectedly have no user mapping even after validation")
+        return
+
+    s = time.monotonic()
+    dbapi_conn = timescale.engine.raw_connection()
+    try:
+        cursor = dbapi_conn.cursor()
+        query = """UPDATE listen AS t SET user_id = e.user_id
+                     FROM (values %%s) AS e (user_name, user_id)
+                    WHERE e.user_name = t.user_name
+                      AND t.listened_at >= %s 
+                      AND t.listened_at < %s"""
+        # Push in chunk boundaries
+        query = cursor.mogrify(query, (chunk_start, chunk_start + CHUNK_SECONDS))
+        psycopg2.extras.execute_values(cursor, query, user_mapping.items(), template=None, page_size=100)
+        dbapi_conn.commit()
+    finally:
+        dbapi_conn.close()
+    e = time.monotonic()
+    print(f" - processed chunk in {round(e-s, 2)}s")
+
+    return True
+
+
+def fill_userid():
+    with timescale.engine.connect() as connection:
+        result = connection.execute(
+            text("""
+               SELECT min(listened_at) as min_listened_at
+                    , max(listened_at) as max_listened_at
+                 FROM listen
+                WHERE listened_at >= :least_accepted_ts
+                  AND user_id = 0"""),
+            least_accepted_ts=LISTEN_MINIMUM_TS)
+        row = result.fetchone()
+        min_listened_at = row['min_listened_at']
+        max_listened_at = row['max_listened_at']
+
+    if min_listened_at is None:
+        print("No chunks found for processing.")
+        return
+
+    number_chunks = math.ceil((max_listened_at - min_listened_at) / CHUNK_SECONDS)
+    chunk_count = 0
+    start_time = time.monotonic()
+    chunk_start = min_listened_at
+    while chunk_start < max_listened_at:
+        result = process_chunk(chunk_start)
+        if result is not True:
+            # If we quit early, then remove this chunk from the count
+            # so that completion time is more accurate
+            number_chunks -= 1
+        else:
+            chunk_count += 1
+        print_status_update(chunk_count, number_chunks, start_time)
+        chunk_start += CHUNK_SECONDS
+
+    print("Finished.")
+    print(f"Usernames who have listens but no known user row: {global_missing_users}")
+
+
+def print_status_update(chunk_count, number_chunks, start_time):
+    """Print a basic status update based on how many chunks have been computed"""
+    chunk_time = time.monotonic()
+    chunk_percentage = chunk_count / number_chunks
```

Comment:
one small bug that I noticed here, if number_chunks is 0 (for example if all chunks are done and there are none remaining) then it's a /0 error. I'd just move this whole calculation into the if block below, and then don't print any time data in the else